lib/streamlit/connections/snowflake_connection.py

```diff
@@ -0,0 +1,260 @@
+        identical to that of using ``@st.cache_data``) as well as simple error handling/retries.
+
+        .. note::
+            Queries that are run without a specified ttl are cached indefinitely.
+
+        Parameters
+        ----------
+        sql : str
+            The read-only SQL query to execute.
+        ttl : float, int, timedelta or None
+            The maximum number of seconds to keep results in the cache, or
+            None if cached results should not expire. The default is None.
+        show_spinner : boolean or string
+            Enable the spinner. The default is to show a spinner when there is a
+            "cache miss" and the cached resource is being created. If a string, the value
+            of the show_spinner param will be used for the spinner text.
+        params : list, tuple, dict or None
+            List of parameters to pass to the execute method. The syntax used to pass
+            parameters is database driver dependent. Check your database driver
+            documentation for which of the five syntax styles, described in `PEP 249
+            paramstyle <https://peps.python.org/pep-0249/#paramstyle>`_, is supported.
+            Default is None.
+
+        Returns
+        -------
+        pd.DataFrame
+            The result of running the query, formatted as a pandas DataFrame.
+
+        Example
+        -------
+        >>> import streamlit as st
+        >>>
+        >>> conn = st.connection("snowflake")
+        >>> df = conn.query("select * from pet_owners")
+        >>> st.dataframe(df)
+        """
+        # TODO(vdonato): Make our error handling more specific if possible. This may be
+        # difficult to do given the limited documentation on the different connector
+        # error subclasses + how many there are.
+        from snowflake.connector import Error as SnowflakeError
+        from tenacity import (
+            retry,
+            retry_if_exception_type,
+            stop_after_attempt,
+            wait_fixed,
+        )
+
+        @retry(
+            after=lambda _: self.reset(),
+            stop=stop_after_attempt(3),
+            reraise=True,
+            retry=retry_if_exception_type(SnowflakeError),
+            wait=wait_fixed(1),
+        )
+        @cache_data(
+            show_spinner=show_spinner,
+            ttl=ttl,
+        )
+        def _query(sql: str) -> pd.DataFrame:
+            cur = self._instance.cursor()
+            cur.execute(sql, **(params or {}))
+            return cur.fetch_pandas_all()
+
+        return _query(sql)
+
+    def write_pandas(
+        self,
+        df: pd.DataFrame,
+        table_name: str,
+        database: str | None = None,
+        schema: str | None = None,
+        chunk_size: int | None = None,
+        **kwargs,
+    ) -> tuple[  # This return type is somewhat horrifying but is the return type of pandas_tools.write_pandas, so we keep it consistent here.
```

Comment:
Pandas doesn't have a type alias for this mess? 😞

Reply:
Doesn't seem like it given this type signature is just copied from the function we're wrapping 😞

Thankfully, it looks like we're going to be simplifying this type in the helper function to just return the first 3 elements of the tuple given how rarely the last element is used, so that simplifies this by a good amount